pyo3-macros-backend/src/pyclass.rs

```diff
@@ -981,26 +980,6 @@ impl<'a> PyClassImplsBuilder<'a> {
             Vec::new()
         }
     }
-
-    /// Enforce at compile time that PyGCProtocol is implemented
-    fn impl_gc(&self) -> TokenStream {
-        let cls = self.cls;
-        let attr = self.attr;
-        if attr.is_gc {
-            let closure_name = format!("__assertion_closure_{}", cls);
-            let closure_token = syn::Ident::new(&closure_name, Span::call_site());
-            quote! {
-                fn #closure_token() {
-                    use _pyo3::class;
-
-                    fn _assert_implements_protocol<'p, T: _pyo3::class::PyGCProtocol<'p>>() {}
-                    _assert_implements_protocol::<#cls>();
-                }
-            }
-        } else {
-            quote! {}
-        }
-    }
```

Comment:
This check isn't needed on Python 3.11, however it _is_ needed on older Python versions else `#[pyclass(gc)]` without `__traverse__` will cause segfaults.

I think a better solution might be to deprecate `#[pyclass(gc)]` and then in `src/pyclass.rs` we can check for the presence of `__traverse__` to decide whether to set the `Py_TPFLAGS_HAVE_GC` flag.